File: src/bridge/protocol.ts
export interface ActiveFile {
  path: string;
  relativePath: string;
  languageId: string;
  lineCount: number;
  cursorLine: number;
  cursorColumn: number;
}

export interface Selection {
  text: string;
  startLine: number;
  startColumn: number;
  endLine: number;
  endColumn: number;
  filePath: string;
}

export interface VisibleFile {
  path: string;
  isActive: boolean;
  isDirty: boolean;
  viewColumn: number;
}

export interface TerminalInfo {
  name: string;
  isActive: boolean;
  hasShellIntegration: boolean;
}

export interface TerminalExecution {
  command: string;
  output: string;
  exitCode: number | undefined;
  cwd: string | undefined;
  timestamp: number;
  terminalName: string;
}

export interface EditorState {
  version: 1;
  timestamp: number;
  pid: number;
  sessionId: string;
  ideName: string;
  workspace: {
    folders: string[];
    name: string;
  };
  activeFile: ActiveFile | null;
  selection: Selection | null;
  visibleFiles: VisibleFile[];
  terminals: TerminalInfo[];
}

export type CommandType = "openFile" | "revealLine" | "getSelection" | "getDiagnostics" | "showDiff" | "previewEdit" | "closePreview" | "getTerminalOutput";

export interface OpenFileArgs {
  path: string;
  line?: number;
  column?: number;
  preview?: boolean;
}

export interface RevealLineArgs {
  path: string;
  line: number;
}

export interface GetSelectionArgs {}

export interface GetDiagnosticsArgs {
  path?: string;
}

export interface ShowDiffArgs {
  path: string;
}

export interface PreviewEditArgs {
  file_path: string;
  tool_name: string;
  old_string?: string;
  new_string?: string;
  content?: string;
}

export interface GetTerminalOutputArgs {
  terminalName?: string;
  limit?: number;
}

export type CommandArgs = OpenFileArgs | RevealLineArgs | GetSelectionArgs | GetDiagnosticsArgs | ShowDiffArgs | PreviewEditArgs | GetTerminalOutputArgs;

export interface BridgeCommand {
  version: 1;
  id: string;
  timestamp: number;
  command: CommandType;
  args: CommandArgs;
}

export interface CommandResult {
  version: 1;
  id: string;
  timestamp: number;
  status: "ok" | "error";
  result: {
    message: string;
    data?: unknown;
  };
}

File: jetbrains-plugin/src/main/kotlin/com/raymondsun24/bridgely/Protocol.kt
package com.raymondsun24.bridgely

data class ActiveFile(
    val path: String,
    val relativePath: String,
    val languageId: String,
    val lineCount: Int,
    val cursorLine: Int,
    val cursorColumn: Int,
)

data class SelectionInfo(
    val text: String,
    val startLine: Int,
    val startColumn: Int,
    val endLine: Int,
    val endColumn: Int,
    val filePath: String,
)

data class VisibleFile(
    val path: String,
    val isActive: Boolean,
    val isDirty: Boolean,
    val viewColumn: Int,
)

data class WorkspaceInfo(
    val folders: List<String>,
    val name: String,
)

data class EditorState(
    val version: Int = 1,
    val timestamp: Long,
    val pid: Long,
    val sessionId: String,
    val ideName: String,
    val workspace: WorkspaceInfo,
    val activeFile: ActiveFile?,
    val selection: SelectionInfo?,
    val visibleFiles: List<VisibleFile>,
)

/** Incoming command written by the MCP server or CLI. */
data class BridgeCommand(
    val version: Int,
    val id: String,
    val timestamp: Long,
    val command: String,
    val args: Map<String, Any?>,
)

data class CommandResultPayload(
    val message: String,
    val data: Any? = null,
)

/** Outgoing result written by the plugin after executing a command. */
data class CommandResult(
    val version: Int = 1,
    val id: String,
    val timestamp: Long,
    val status: String,   // "ok" | "error"
    val result: CommandResultPayload,
)
